### PR TITLE
If entry data is changed, file name should also be changed.

### LINKED ISSUE #11316: If entry data is changed, file name should also be changed
- An entry has a file attached
- The filename matches "rename to the defined pattern" (meaning auto generated file name; configured in Preferences -> Linked files -> Linked file name conventions)
- I change the data also changing the name generated the defined pattern
- JabRef should rename the file, too

Tasks:

1. Add a preference Linked files -> Linked file name conventions -> `[ ] Auto rename files if entry changes` (false as default)
2. Write test cases
3. Listen to entry change events and check preference and rename file (if preference allows and file name matches)
Note that an empty pattern should not result in any action (because the file name does not match)

---

Important: Also consider the case that the entry has multiple files attached.

### PR DESCRIPTION
**Pull request by AdiAr11**
Closes #11316 

### Mandatory checks

- [x] I own the copyright of the code submitted and I licence it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

UI change:
Screenshot 2024-10-22 at 7 22 24 PM

### CODE REVIEW

**Comment by AdiAr11** [comment]
Screenshot 2024-10-22 at 7 22 24 PM

**Comment by AdiAr11** [comment]
Hello, @HoussemNasri, good evening. I have implemented the changes. Please let me know if any changes are needed.
Thank you!

File: src/main/java/org/jabref/gui/externalfiles/LinkedFileAutoRenamer.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    public void listen(FieldChangedEvent event) {
+        if (!preferences.shouldAutoRenameLinkedFiles()) {
+            return;
+        }
+
+        BibEntry entry = event.getBibEntry();
+        Field changedField = event.getField();
+
+        String fileNamePattern = preferences.getFilePreferences().getFileNamePattern();
+        if (fileNamePattern == null || fileNamePattern.trim().isEmpty()) {
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/gui/externalfiles/LinkedFileAutoRenamer.java:44]: isBlank() instead of trim().isEmpty()

Comment (AdiAr11) [reply src/main/java/org/jabref/gui/externalfiles/LinkedFileAutoRenamer.java]: Made this change as suggested by you!

Comment (AdiAr11) [reply src/main/java/org/jabref/gui/externalfiles/LinkedFileAutoRenamer.java]: @HoussemNasri Any other changes?

File: src/main/java/org/jabref/logic/util/io/FileUtil.java [hunk 2]
```diff
@@ -524,4 +524,66 @@ public static boolean detectBadFileName(String fileName) {
     public static boolean isCharLegal(char c) {
         return Arrays.binarySearch(ILLEGAL_CHARS, c) < 0;
     }
+
+    public static boolean renameLinkedFile(LinkedFile linkedFile, BibEntry entry, String fileNamePattern, List<Path> fileDirectories, FilePreferences filePreferences, BibDatabaseContext bibDatabaseContext) {
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/logic/util/io/FileUtil.java:528]: Is this method duplicated from somewhere else in the codebase? I'm sure we have already implemented this.

File: src/main/java/org/jabref/gui/externalfiles/LinkedFileAutoRenamer.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        LOGGER.debug("Handling entry change for entry: {}", entry.getCitationKey().orElse("unknown"));
+
+        List<LinkedFile> linkedFiles = entry.getFiles();
+
+        for (LinkedFile linkedFile : linkedFiles) {
+            boolean renamed = FileUtil.renameLinkedFile(linkedFile, entry, fileNamePattern, fileDirectories, preferences.getFilePreferences(), bibDatabaseContext);
+
+            if (renamed) {
+                UiTaskExecutor.runInJavaFXThread(() -> {
+                    entry.setFiles(linkedFiles);
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/gui/externalfiles/LinkedFileAutoRenamer.java:83]: I'm uncertain whether we need to do this for every linked file. The typical user would have a handful or linked files per entry, but you never now, people can get very creative, so if it's possible to minimize calls to setFiles, we should do it.

File: src/main/java/org/jabref/logic/util/io/FileUtil.java (hunk 2 above)
Comment (HoussemNasri) [reply src/main/java/org/jabref/logic/util/io/FileUtil.java:528]: We have another implementation at `LinkedFileHandler` that differs slightly from this one. Can we refactor this to remove the duplication? Perhaps we can combine both implementations and change dependents of LinkedFileHandler and LinkedFileAutoRenamer to depend on the same implementation.

File: src/main/java/org/jabref/gui/externalfiles/FileNameParser.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package org.jabref.gui.externalfiles;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class FileNameParser {
+
+    private static final Pattern FIELD_PATTERN = Pattern.compile("\\[([^\\]]+)\\]");
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/gui/externalfiles/FileNameParser.java:9]: I'm pretty sure this is already implemented. Just to clarify, it is possible to change the linked file name to follow a pattern by right-clicking on the linked file and choosing 'rename file to a defined pattern'.

**Comment by AdiAr11** [comment]
@HoussemNasri Thank you so much for the feedback. I will try my best to make these changes as soon as possible!

File: src/main/java/org/jabref/gui/externalfiles/LinkedFileAutoRenamer.java (hunk 3 above)
Comment (AdiAr11) [reply src/main/java/org/jabref/gui/externalfiles/LinkedFileAutoRenamer.java:83]: Hi @HoussemNasri, do you have any suggestion on what should be done for this?

File: src/main/java/org/jabref/gui/externalfiles/FileNameParser.java (hunk 4 above)
Comment (AdiAr11) [reply src/main/java/org/jabref/gui/externalfiles/FileNameParser.java]: Changed the code to use BracketedPattern instead and deleted this class!!

File: src/main/java/org/jabref/logic/util/io/FileUtil.java (hunk 2 above)
Comment (AdiAr11) [reply src/main/java/org/jabref/logic/util/io/FileUtil.java]: Did this as well!

**Comment by AdiAr11** [comment]
Hi @HoussemNasri, i have made the changes you requested. Hope it is fine now.
Thank you!

File: src/main/java/org/jabref/gui/externalfiles/LinkedFileAutoRenamer.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+                if (renamed) {
+                    UiTaskExecutor.runInJavaFXThread(() -> {
+                        entry.setFiles(linkedFiles);
+                        LOGGER.debug("Updated entry files after renaming.");
+                    });
+                }
+            } catch (IOException e) {
+                LOGGER.error("Failed to rename file '{}' to '{}'", linkedFile.getLink(), fileNamePattern, e);
+            }
+        }
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/gui/externalfiles/LinkedFileAutoRenamer.java:105]:
```suggestion
                FileUtil.renameLinkedFileToName(linkedFile, entry, newFileName, false, bibDatabaseContext, fileDirectories, preferences.getFilePreferences());
                }
            } catch (IOException e) {
                LOGGER.error("Failed to rename file '{}' to '{}'", linkedFile.getLink(), fileNamePattern, e);
            }
        }
        entry.setFiles(linkedFiles);
        LOGGER.debug("Updated entry files after renaming.");
```

Scheduling the linked files update on the JavaFX thread will lead to glitches if you change the fields fast enough. Also, we only need to call `setFiles` once we have renamed all the linked files that needed renaming.

File: src/main/java/org/jabref/gui/externalfiles/LinkedFileAutoRenamer.java (hunk 3 above)
Comment (HoussemNasri) [reply src/main/java/org/jabref/gui/externalfiles/LinkedFileAutoRenamer.java]: Please checkout the comment below.

File: src/main/java/org/jabref/gui/externalfiles/LinkedFileAutoRenamer.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        LOGGER.debug("Handling entry change for entry: {}", entry.getCitationKey().orElse("unknown"));
+
+        List<LinkedFile> linkedFiles = entry.getFiles();
+
+        for (LinkedFile linkedFile : linkedFiles) {
+            try {
+                String newFileName = FileUtil.createFileNameFromPattern(bibDatabaseContext.getDatabase(), entry, fileNamePattern);
+                LOGGER.debug("Generated new file name: '{}'", newFileName);
+
+                boolean renamed = FileUtil.renameLinkedFileToName(linkedFile, entry, newFileName, false, bibDatabaseContext, fileDirectories, preferences.getFilePreferences());
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/gui/externalfiles/LinkedFileAutoRenamer.java:94]: Passing the boolean `false` like this is against best practices, please define a constant OVEWRITE_EXISTING_FILE = true, and pass `!OVEWRITE_EXISTING_FILE` to the method.

File: src/main/java/org/jabref/logic/util/io/FileUtil.java [hunk 7]
```diff
@@ -588,4 +588,73 @@ public static String shortenFileName(String fileName, Integer maxLength) {
     public static boolean isCharLegal(char c) {
         return Arrays.binarySearch(ILLEGAL_CHARS, c) < 0;
     }
+
+    public static boolean renameLinkedFileToName(LinkedFile linkedFile,
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/logic/util/io/FileUtil.java:592]: This is too much logic to go untested, please unit-test this.

**Comment by AdiAr11** [comment]
Noted! Will make these changes and write the test cases. Thank you!

File: src/main/java/org/jabref/logic/util/io/FileUtil.java (hunk 7 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/logic/util/io/FileUtil.java:592]: You can call the already existing RenamePdfCleanup which does exactly this

**Comment by koppor** [comment]
Closing this issue due to inactivity :zzz: 

We discussed in the DevCall. The implementation should make use of the existing renaming cleanup.

**Pull request closed without merging**